### DISCUSSION TITLE
Simple way to use a DAQ device with two AOs?

### DISCUSSION THREAD

**Discussion by syim70** [Ulm-IQO/qudi-iqo-modules#146] [Q&A] [answered]
Thanks for all your hard work to build qudi. Maybe I can't contribute much except to say thank you.

I am trying to run the 'scanner' of qudi with my NI DAQ device, but fails because it has only two analog outputs. I understand 'scanner' requires at least three AOs (my NI DAQ works well with 'ni_x_series_in_streamer').

Is there a simple way to fix this? I tried to modify the example 'config' and run it, as I I don't need z-axis control at present (x-z scan, y-z scan, optimizer, etc.), but it failed.

If the modification of 'config' is not enough, would you recommend what I should do?

Thank you in advance.

**Comment by Neverhorst** [comment, marked as answer]
Hi @syim70 ,

a "thank you" is much appreciated. :)

Are you referring to this error upon activating the `ScannerGui` with dummy config and only 2 scan axes configured?
```
Traceback (most recent call last):
  File "C:\Software\qudi-core\src\qudi\core\module.py", line 301, in __activation_callback
    self.on_activate()
  File "C:\Software\qudi-iqo-modules\src\qudi\gui\scanning\scannergui.py", line 182, in on_activate
    self._init_optimizer_settings()
  File "C:\Software\qudi-iqo-modules\src\qudi\gui\scanning\scannergui.py", line 349, in _init_optimizer_settings
    self.update_optimizer_settings()
  File "C:\Software\qudi-iqo-modules\src\qudi\gui\scanning\scannergui.py", line 1019, in update_optimizer_settings
    new_settings = self._optimize_logic().check_sanity_optimizer_settings(settings, self._optimizer_plot_dims)
  File "C:\Software\qudi-iqo-modules\src\qudi\logic\scanning_optimize_logic.py", line 231, in check_sanity_optimizer_settings
    raise ValueError(f"Configured optimizer dim= {plot_dimensions}"
ValueError: Configured optimizer dim= [2, 1] doesn't yield any sensible scan sequence.
```
That is a result from the GUI trying to create a scan widget for an [xy, z]-optimization sequence. 
If you specify the following option in your configuration for the `ScannerGui` everything should work and you should get an optimizer scan widget with a xy-scan:
``` yaml
optimizer_plot_dimensions: [2]
```
See also [This example config for `ScannerGui`](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/622b43a71ead7eb8dddf8e8b226c421aa8d2987a/src/qudi/gui/scanning/scannergui.py#L95)

However one needs to delve quite deep into the code to find this solution and it is not very obvious from the error message what happens. So we should improve the error handling here.

Thank you for bringing this up.

**Reply by syim70** [reply]
Dear Neverhorst,

I have tested my system as you advised, and it works well!
For many people, it may be an easy problem, but It was a big huddle for me.
Thank you so much for the kind reply:)

**Reply by timoML** [reply]
#155 will improve the handling with non standard scanner/optimizer configurations. Maybe you could test, whether this works smoother wird your 2d configuration..?

**Reply by syim70** [reply]
Dear timoML,
Thank you for your commitment to improvement. It works well for my application!

**Comment by syim70** [comment]
Dear Neverhorst and timoML,

After I have upgraded the iqo-module due to Numpy 2.0 issue, I have a new problem with my scanner.
So, I tried many ways, and finally found that the error came up even with scanner_dummy itself.
It works well with "optimizer_plot_dimensions: [2, 1]" in config file, but not with "optimizer_plot_dimensions: [2]".

I am using Release v0.5.1 qudi-iqo-module, and Release v1.5.1. of qudi-core.
Thank you for your support.

**Reply by Neverhorst** [reply]
Hi @syim70 ,

since this topic seems to be unrelated to your original question above, would you please [open a bug report here](https://github.com/Ulm-IQO/qudi-iqo-modules/issues/new/choose) and provide more details on the error you are encountering? Thank you.
I will lock this conversation for now since your original question has been answered.

**Reply by timoML** [reply]
https://github.com/Ulm-IQO/qudi-iqo-modules/pull/155 is also reworking the dummy. Checking out this branch should solve both of your issues. We are in the process of testing this PR and will merge this soonish.
